Track Quality Integration

Chriisbrown opened the pull request:
#### PR description:

Addition of ML based track quality classifiers, includes the track quality class that performs a track quality prediction using an ONNX runtime when called in the track finding. It will fill the track MVA1 variable which can be outputted into the L1TrackNtuples and can then be analysed with the L1TrackQualityPlot.C root macro.
By default this track quality addition is switched off but is activated by the TrackQuality flag in the Tracklet_cfi file, to output to Ntuples the TrackQuality flag in the L1TrackNtupleMaker_cfg also needs to be set to true.

To run this code it needs an additional dependency of a ONNX model file located in L1Trigger/TrackTrigger/data, this file will be added to the cms-data repository under cms-data/L1Trigger-TrackTrigger but for now a default model can be added to run this quality code by running: git clone https://github.com/ChriisBrown/L1Trigger-TrackTrigger.git L1Trigger/TrackTrigger/data

The talk relating to this PR is found here: https://indico.cern.ch/event/954660/contributions/4018027/attachments/2103671/3537386/L1TrackQuality.pdf

#### PR validation:

Ran all checks found here: https://twiki.cern.ch/twiki/bin/view/CMS/L1TrackCodeRules
Ran separate tests using the L1TrackNtupleMaker_cfg and HybridTracks_cfg to check ONNX model output vs python predictions
Ran scram b runtests with no issues

Chriisbrown committed:
Track Quality Integration

Chriisbrown committed:
Track Ntuple Maker Max Events

cgsavard commented on L1Trigger/TrackFindingTracklet/test/L1TrackNtupleMaker_cfg.py:
```diff
@@ -169,6 +169,8 @@ def getTxtFile(txtFileName):
                                        TrackingInJets = cms.bool(False),
                                        GenJetInputTag = cms.InputTag("ak4GenJets", ""),
+                                       # Was track quality performed?
+                                       TrackQuality = cms.bool(False)
```
If the current default model is not included, then perhaps we should include a comment here with a link to where to find this model. Something like "Is track quality performed? (--> requires model in L1Trigger/TrackTrigger/data present! default: https://github.com/Chriisbrown/L1Trigger-TrackTrigger)". Is there a better place to include this link?

Chriisbrown replied:
I have a  comment in the TrackQualityParams_cfi.py file but might also be worth having the same comment in the Ntuple maker too?

cgsavard reviewed (commented):
This looks great, thanks Chris! My only comment is that I think it currently is not obvious to the user that they need to include a model to make this run. I put a comment in where I think might be a good place to let the user know. Perhaps there is a better place or another way to do this?

tomalin reviewed (commented):
I would expect that people running with the default cfg, which disables the track quality code with "TrackQuality=cms.bool(False)" in Tracklet_cfi.py, wouldn't need to download your your GBDT_model.onnx file. However, if I try running like this, the code throws a fatal error.
Perhaps L1FPGATrackProducer should only instantiate the TrackQuality object (e.g. via a smart pointer) if TrackQuality=True?

tomalin reviewed (commented):
I suggest removing cfg parameter "TrackQuality" from L1TrackNtupleMaker, to save people the pain of forgetting to set its value correctly. If we leave it permanently True, it doesn't increase the size of the ntuple much. And since TTTrack assigns a default value of 0 to MVA if the L1 tracking doesn't run Track Quality, L1TrackNtupleMaker doesn't need to set it to -999 if TrackQuality=False.

Also, why does L1TrackNtupleMaker.cc create a branch m_matchtrk_chi2, but not m_matchtrk_MVA?

cgsavard commented:
> However, if I try running like this, the code throws a fatal error.

Ah yes, I do also get this error when they model is not loaded in. So at the moment, the user needs the model regardless of whether they want to run track quality or not. I agree with only instantiating the object in TrackQuality=True.

> If we leave it permanently True, it doesn't increase the size of the ntuple much.

I'm not sure that we want to default this to True at the moment. If it is True, then it would be ideal to provide the default model somewhere in the framework so that it is not extra work to get the whole ntuple running out of the box. Is there any way to put the model in cmssw? Or can we set up some code that runs by default that clones the model to the right repository perhaps?

I also think it should default to False because the default model is not necessarily the best model. I don't want people to use this model thinking that it is the current model everyone should use when I think of it more as an example model.

tomalin commented:
> I'm not sure that we want to default this to True at the moment. If it is True, then it would be ideal to provide the default model somewhere in the framework so that it is not extra work to get the whole ntuple running out of the box. Is there any way to put the model in cmssw? Or can we set up some code that runs by default that clones the model to the right repository perhaps? I also think it should default to False because the default model is not necessarily the best model. I don't want people to use this model thinking that it is the current model everyone should use when I think of it more as an example model.

A: I think there may be confusion here. There are two distinct cfg params named "TrackQuality", one appearing in Tracklet_cfi.py, which should remain there, and the other appearing at https://github.com/Chriisbrown/cmssw/blob/L1TK-dev-11_2_0_pre6/L1Trigger/TrackFindingTracklet/test/L1TrackNtupleMaker_cfg.py#L173, which I propose removing, with L1TrackNtupleMaker.cc then always creating a branch in its output TTree for "MVA". This change would not prevent the user setting TrackQuality=False in Tracklet_cfi.py, if they chose, and running the L1 tracking without an onnx model.

tomalin reviewed (commented):
1) Using "gSystem->mkdir()" inside L1TrackQualityPlot.C was a good idea. Please add a gSystem->mkdir("TrkPlots") to L1TrackNtuplePlot.C
2) Please add comments to top of L1TrackQualityPlot.C explaining obscure variables in plots. e.g. "TPR=Efficiency, FPR=Fake Rate, dt=cut on MVA variable, ROC" etc. (Though I assume TPR is not efficiency, since it goes to 100%).
3) For sake of colour blind people, where plots use two colours, please make them blue & green, and make curves a little thicker ...
4) Would it be a good idea to merge L1TrackQualityPlot.C and L1TrackNtuplePlot.C, or is it best to keep them separate? Just asking ...

Chriisbrown committed:
PR Code Review Fixes #1

Chriisbrown commented:
I have updated the code to reflect the comments made, major changes are:
-Updated the way the TrackQuality model and the ONNX file are loaded so you can run the tracking without needing an ONNX model by default.
-Removed the TrackQuality flag in the Ntuple maker python config and analyser file and added a matchtrk_MVA1 branch, by default both the trk_MVA1 and matchtrk_MVA1 are filled with 0s when the ONNX models are not used (question over whether this is confusing seeing as when the ONNX models are run 0 is used as a "fake" classification)

cgsavard commented:
> 2. Please add comments to top of L1TrackQualityPlot.C explaining obscure variables in plots. e.g. "TPR=Efficiency, FPR=Fake Rate, dt=cut on MVA variable, ROC" etc. (Though I assume TPR is not efficiency, since it goes to 100%).

Right, it is not quite efficiency as it is applied to the reco tracks and trying to imitate the matching. It should instead be "TPR = fraction of "real" L1 tracks that were correctly identified, FPR = fraction of "fake" L1 tracks that were incorrectly identified,  dt = cut on MVA variable that determines "real" or "fake""

> 3. For sake of colour blind people, where plots use two colours, please make them blue & green, and make curves a little thicker ...

Yes, good point.

> 4. Would it be a good idea to merge L1TrackQualityPlot.C and L1TrackNtuplePlot.C, or is it best to keep them separate? Just asking ...

I was unsure about this. If they are merged then there should be a flag to run this portion in case TrackQuality=False. On one hand, it's nice to have all the plots in one place. On the other hand, I can imagine people wanting to use only the MVA plots as the model to use is not set and can be configured by the user. I have a slight preference to keep them separate for this reason.

Chriisbrown committed:
Comments and style changes

tomalin reviewed (approved):
Looks finished to me.
(Though at some point, suggest replacing SetLineColor(3) by SetLineColor(kGreen) etc. for better comprehensibility. See https://root.cern.ch/doc/master/classTColor.html )
